go/arrow/array/dictionary.go

```diff
@@ -0,0 +1,1305 @@
+// an ArrayData object with a datatype of arrow.Dictionary and a dictionary
+func NewDictionaryData(data arrow.ArrayData) *Dictionary {
+	a := &Dictionary{}
+	a.refCount = 1
+	a.setData(data.(*Data))
+	return a
+}
+
+func (d *Dictionary) Retain() {
+	atomic.AddInt64(&d.refCount, 1)
+}
+
+func (d *Dictionary) Release() {
+	debug.Assert(atomic.LoadInt64(&d.refCount) > 0, "too many releases")
+
+	if atomic.AddInt64(&d.refCount, -1) == 0 {
+		d.data.Release()
+		d.data, d.nullBitmapBytes = nil, nil
+		d.indices.Release()
+		d.indices = nil
+		if d.dict != nil {
+			d.dict.Release()
+			d.dict = nil
+		}
+	}
+}
+
+func (d *Dictionary) setData(data *Data) {
+	d.array.setData(data)
+
+	if data.dictionary == nil {
+		panic("arrow/array: no dictionary set in Data for Dictionary array")
+	}
+
+	dictType := data.dtype.(*arrow.DictionaryType)
+	debug.Assert(arrow.TypeEqual(dictType.ValueType, data.dictionary.DataType()), "mismatched dictionary value types")
+
+	indexData := NewData(dictType.IndexType, data.length, data.buffers, data.childData, data.nulls, data.offset)
+	defer indexData.Release()
+	d.indices = MakeFromData(indexData)
+}
+
+// Dictionary returns the values array that makes up the dictionary for this
+// array.
+func (d *Dictionary) Dictionary() Interface {
+	if d.dict == nil {
+		d.dict = MakeFromData(d.data.dictionary)
+	}
+	return d.dict
+}
+
+// Indices returns the underlying array of indices as it's own array
+func (d *Dictionary) Indices() Interface {
+	return d.indices
+}
+
+// CanCompareIndices returns true if the dictionary arrays can be compared
+// without having to unify the dictionaries themselves first.
+func (d *Dictionary) CanCompareIndices(other *Dictionary) bool {
+	if !arrow.TypeEqual(d.indices.DataType(), other.indices.DataType()) {
+		return false
+	}
+
+	minlen := int64(min(d.data.dictionary.length, other.data.dictionary.length))
+	return ArraySliceEqual(d.Dictionary(), 0, minlen, other.Dictionary(), 0, minlen)
+}
+
+func (d *Dictionary) String() string {
+	return fmt.Sprintf("{ dictionary: %v\n  indices: %v }", d.Dictionary(), d.Indices())
+}
+
+// GetValueIndex returns the dictionary index for the value at index i of the array.
+// The actual value can be retrieved by using d.Dictionary().(valuetype).Value(d.GetValueIndex(i))
+func (d *Dictionary) GetValueIndex(i int) int {
+	indiceData := d.data.buffers[1].Bytes()
+	// we know the value is non-negative per the spec, so
+	// we can use the unsigned value regardless.
+	switch d.indices.DataType().ID() {
+	case arrow.UINT8, arrow.INT8:
+		return int(uint8(indiceData[d.data.offset+i]))
+	case arrow.UINT16, arrow.INT16:
+		return int(arrow.Uint16Traits.CastFromBytes(indiceData)[d.data.offset+i])
+	case arrow.UINT32, arrow.INT32:
+		return int(arrow.Uint32Traits.CastFromBytes(indiceData)[d.data.offset+i])
+	case arrow.UINT64, arrow.INT64:
+		return int(arrow.Uint64Traits.CastFromBytes(indiceData)[d.data.offset+i])
```

Comment:
unlikely to hit this in practice but doesn't this potentially truncate the index?

Reply:
fair point. On a 64-bit machine if you have uint64 indexes and the value is larger than can fit in an int64 it would truncate. 

On a 32-bit machine this'll truncate for any 64-bit index type and for uint32 indexes that are larger than an int32. I guess I should change this to return a uint64 instead? so that we know we never truncate? What do you think?

Reply:
maybe just panic for truncation?  like I said i think this is probably unlikely.  If you want to go through the effort int64 seems sufficient.  most architectures I'm aware of can't even address a full 64 bits of ram.

Reply:
Sure, i'll add a check for truncation and panic.

Reply:
Actually, given how unlikely this, I'm going to make it a `debug.Assert` to ensure that it doesn't affect performance unless compiled with the asserts turned on.